### DISCUSSION TITLE
Anonymous or pseudo-anonymous user-id

### DISCUSSION THREAD

**Discussion by kristianrunelarsen** [plausible/analytics#601] [Anything Else]
Hi 

We want to use plausible but I am not certain that the generated user-id is anonymous in terms of the GDPR or if it is considered as pseudo-anonymous. As I understand it all-hashing is considered equal in terms of GDPR and is considered as pseudo-anonymization. 

![Screenshot from 2021-01-14 10-01-23](https://user-images.githubusercontent.com/67627991/104568118-f2275080-564f-11eb-955c-5b61a117b891.png)

In terms of the pseudo-anonymous/anonymous discussion would it be a fix to remove some info from the user agent string and some digits from the ip before hashing ?

refs: 
* https://academic.oup.com/idpl/article/10/1/11/5802594
* (https://www.kevel.co/blog/gdpr-server/)

**Comment by kristianrunelarsen** [comment]
Just a thought. 

The ip is just there to kind of uniquely identify a user. If we strip some digits from the ip it would result in more clashes, but if the ip was shuffeled with a daily seed all the digits could still be used. I think a potential future "GPDR" brute-force attacker would have a hard time figuring out the real ip because the shuffeled ip would look real. The salt is already generated on a daily basis so the code is almost in place already.

Cheers
Rune

**Reply by ukutaht** [reply]
Thanks for the suggestion @kristianrunelarsen.

To explain the thinking behind the current implementation first. Let's say our database leaked along with the user IDs. The main risk is that someone could identify the original IP address. A rainbow attack is certainly possible because there are only 4.3b IPv4 addresses in the wild and maybe hundreds of common user agent values.

First, we add the `site_id` into the hash so in case anything leaks, an attacker will have a harder time correlating traffic from one IP address on multiple sites.

Second, our strategy for full anonymisation is to inject noise into the hash. We use OpenSSL's `RAND_bytes` method to generate a 128-bit random number. The amount of possible unique values for 128 bits is huge and a rainbow attack on the resulting hash is not computationally realistic without knowing the secret salt.

We consider our database anonymised because the salt is discarded and re-generated every day at midnight. By completely forgetting the random salt that was used, we ensure that even we are not able to reverse the hash function. We 100% rely on key security to provide anonymity.

The EU Data Protection Working Party's [article on anonymisation techinques](https://ec.europa.eu/justice/article-29/documentation/opinion-recommendation/files/2014/wp216_en.pdf) also iterates the importance of key security when using hashing for anonymisation. My interpretation of the following paragraph is that in hashing/encryption, the data is only as secure as the encryption key/hash salt.

> Neither  encryption  nor  key-coding  per  se  lends itself  to  the goal  of  making  a  data  subject unidentifiable: as, in the hands of the controller at least, the original data are still available or deducible. The  sole  implementation  of  a  semantic  translation  of  personal  data,  as  happens with  key-coding,  does  not  eliminate  the  possibility  to restore the  data  back  to  their  original structure -either  by  applying  the  algorithm  in  the  opposite  way,  or  by  brute  force  attacks, depending  on  the  nature  of  the  schemes,  or  as  a result  of  a  data  breach.  State-of-the-art encryption can ensure  that data is protected to  a  higher  degree, i.e.it  is unintelligible  for entities that ignore the decryption key, but it does not necessarily result in anonymisation. For as long as the key or the original data are available (even in the case of a trusted third party, contractually  bound  to  provide  secure  key  escrow  service),  the  possibility  to  identify  a  data subject is not eliminated. . 


So this paragraph stresses the fact that as long as your secret key is somewhere around, a hash/encryption function can be reversed. This is exactly why we don't keep the key around and we throw it away. We believe that by forgetting the hash, it becomes impossible for anyone, including ourselves, with the help of anyone, to identify the subject of the data

Now to your suggestion to shuffle the IP address with a daily seed. I feel like that's pretty much what we're doing already?

If I understand correctly, your suggestion is to truncate the last byte from the IP address before hashing. The effect of this would be to reduce the accuracy of our tracking (lots more hash collisions). I also see the benefit because then all of this hashing and key security would not matter. First 3 bytes of an IP address are not personally identifiable to anyone, therefore not personal data.

We haven't tested how this would affect unique visitor tracking so I can't offer a good opinion on the utility of this method. Is it 10% less accurate, 70%? No idea.

For now, we are comfortable with our anonymisation technique. There is a little bit of trust involved since it's impossible for you to verify that a daily salt hasn't leaked. However, this is why our data processing agreement states the following:
> Plausible Analytics shall implement and maintain appropriate technical and organisational security measures designed to protect the visitor data against unauthorised or unlawful processing and against accidental loss, destruction, damage, theft, alteration or disclosure.

**Reply by ukutaht** [reply]
I think the paper you linked is very valuable in this discussion: https://academic.oup.com/idpl/article/10/1/11/5802594

From there I found more discussion that seems to support proper hashing as an anonymisation techinque. I also wasn't aware of the difference between salt and pepper. Turns out we use pepper not salt for our hashes. Anyways: 

> As long as the pepper is chosen randomly and held secret, the same input data (the password) does not result in the same hash across databases. More importantly, it also hinders brute-force attacks on leaked databases of hashed passwords based on a list (‘base-population’) of popular passwords as any attacker would have to try out all possible peppers for any guessed password. A randomly chosen pepper of 4 bytes (32 bits)—the minimum length suggested by the NIST—increases the necessary guessing effort by the factor of 232 (more than 4 billion) as long as the pepper remains secret.

The paper does stress the importance of keeping the pepper secret and choosing the proper hash function. We need to look at SipHash and whether we should replace this with something cryptographically stronger.

**Reply by kristianrunelarsen** [reply]
I think this whole discussion of anonymization/pseudo-anonymization is a minefield :-). In practice I think that hashing with salt/pepper is sufficient. 

Concerning SipHash I found this: https://crypto.stackexchange.com/questions/17996/is-siphash-cryptographically-secure

I think with hashing, the salt is still in the hash and it can be brute-forced to a certain extend. 

The reason why I suggested to shuffle the IP was that the output would also look like a real IP. I think it would be hard for a potential attacker to figure out the mapping, and you would keep the input length.  If we shuffled the entire string my guess is that you could figure out the mapping because you could guess some of the original ingredients, IP, user-agent, domain. By guessing some of the original ingredients you could make a test and then somehow do a brute force search. In the IP-scenario it would be hard to make such a test.

**Reply by kristianrunelarsen** [reply]
Maybe if you knew you visited the site your self ... Then you would know the real IP and user-agent

**Reply by ukutaht** [reply]
> Concerning SipHash I found this: https://crypto.stackexchange.com/questions/17996/is-siphash-cryptographically-secure
> I think with hashing, the salt is still in the hash and it can be brute-forced to a certain extend.

Right so the issue is that SipHash is 64-bit and therefore finding collisions is doable. But, even if you find ingredients that create a colliding hash, how can you be sure that these were the exact ingredients that actually created the original hash? It could be that a different random salt and IP address result in the same hash value.

So, to confidently find the IP address from our hash you have to try every single possible salt. Our 128-bit salt has `2^128 == 3.4 x 10^38` possible values which is computationally unrealistic to brute force.

> Maybe if you knew you visited the site your self ... Then you would know the real IP and user-agent

Yeah this is the worst case scenario for an attack where someone already knows the IP address and User-Agent to search for. However, the same brute force issues apply as I explained above.

> I think this whole discussion of anonymization/pseudo-anonymization is a minefield :-).

Absolutely. We will definitely talk to lawyers to form a stance on this which is backed by official documentation and references. In absence of case law precedent, this is probably the best we can do at the moment.

**Reply by ukutaht** [reply]
To add to this, we can totally move to a 128-bit hash to make it computationally unrealistic to even find any hash collisions at all. I don't see any problems with this other than increasing our CPU and storage resource consumption marginally.

**Reply by kristianrunelarsen** [reply]
Hi 

Sorry for the late reply.

I have been in contact with a guy in our security department. He liked the method, but he had no experience with siphash so he didn't feel confident about approving it as it is.  He suggested that for it to be completely anonymous we could rotate all the id's the following day by assigning a completely random number for each session id and throw away the old session id. Would that be an easy task in the clickhouse db ?

This could also be done runtime. By setting up a redis cache and have it assign a random value to each session fingerprint. The random value should have et an expiry time at midnight. And then use the random value as fingerprint. 

Cheers
Rune

**Reply by ukutaht** [reply]
ClickhouseDB is hyper-optimized for analytics workloads but one invariant that helps it achieve the blazing fast performance is a presumption of data immutability. I don't think mutating the session ID after it's been collected is a viable option.

Doing something at runtime is a good option. We could keep a mapping from `session_fingerprint -> random_session_id` in memory and just clear the mappings each midnight. In that case, since the fingerprint is never persisted, I don't know if a daily salt rotation is even needed.

The immediate problem I see with this is the need for an external in-memory cache like Redis. We try to keep our infrastructure as small and simple as possible for self-hosters. I've been working hard to not involve another piece like Redis into the mix. The BEAM ecosystem could have something for this use-case, I'll see if mnesia could work for this.

I really like the idea of mapping to random session_ids at runtime.

I do have to say I'm swamped at the moment and I don't think I have time to work on this anytime soon. It feels like a pretty complex piece of work to be honest and I cannot promise any timelines.
